app/models/instance.rb

```diff
@@ -1,19 +1,57 @@
 # frozen_string_literal: true
+# == Schema Information
+#
+# Table name: instances
+#
+#  domain :string           primary key
+#
 
-class Instance
-  include ActiveModel::Model
+class Instance < ApplicationRecord
+  self.primary_key = :domain
 
-  attr_accessor :domain, :accounts_count, :domain_block
+  belongs_to :domain_block, foreign_key: :domain, primary_key: :domain
+  belongs_to :domain_allow, foreign_key: :domain, primary_key: :domain
 
-  def initialize(resource)
-    @domain         = resource.domain
-    @accounts_count = resource.respond_to?(:accounts_count) ? resource.accounts_count : nil
-    @domain_block   = resource.is_a?(DomainBlock) ? resource : DomainBlock.rule_for(domain)
-    @domain_allow   = resource.is_a?(DomainAllow) ? resource : DomainAllow.rule_for(domain)
+  scope :matches_domain, ->(value) { where(arel_table[:domain].matches("%#{value}%")) }
+
+  def readonly?
+    true
+  end
+
+  def delivery_failure_tracker
+    @delivery_failure_tracker ||= DeliveryFailureTracker.new(domain)
+  end
+
+  def accounts_count
+    cache('accounts_count') { accounts_scope.count }
+  end
+
+  def following_count
+    cache('following_count') { Follow.where(account: accounts_scope).count }
+  end
+
+  def followers_count
+    cache('followers_count') { Follow.where(target_account: accounts_scope).count }
+  end
+
+  def reports_count
+    cache('reports_count') { Report.where(target_account: accounts_scope).count }
   end
 
-  def countable?
-    @accounts_count.present?
+  def blocks_count
+    cache('blocks_count') { Block.where(target_account: accounts_scope).count }
+  end
```

Comment:
I am unsure how useful that caching is, and there is no explicit invalidation mechanism nor any expiration time set, so it is very likely to become stale.

Reply:
Default cache expiration time is 10 minutes

Reply:
ok, maybe making it explicit would be better, but it's fine